lib/src/Base/Type/MatrixImplementation.cxx

```diff
@@ -1620,5 +1620,142 @@ UnsignedInteger MatrixImplementation::stride(const UnsignedInteger dim) const
   return stride;
 }
 
+/** Diagonal extraction */
+MatrixImplementation MatrixImplementation::getDiagonal(const SignedInteger k) const
+{
+
+  if (k >= nbColumns_) throw OutOfBoundException(HERE) << "One must have k < nbColumns";
+  if (-k > nbRows_) throw OutOfBoundException(HERE) << "One must have -nbRows < k ";
```

Comment:
This does not seem to work for an IdentityMatrix. The following code raises an exception:

```
mat = ot.IdentityMatrix(5)
mat.getDiagonal(1)
```